<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Release history &mdash; CPAT 3.0.5 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b76e3c8a" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=d6b40703"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            CPAT
          </a>
              <div class="version">
                3.0.5
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Release history</a></li>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#prerequisite">Prerequisite</a></li>
<li><a class="reference internal" href="#install-cpat-using-pip3">install CPAT using pip3</a></li>
</ul>
</li>
<li><a class="reference internal" href="#run-cpat-online">Run CPAT online</a></li>
<li><a class="reference internal" href="#run-cpat-on-local-computer">Run CPAT on local computer</a><ul>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#build-your-own-hexamer-table">Build your own hexamer table</a><ul>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#id3">Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#build-your-own-logit-model">Build your own logit model</a><ul>
<li><a class="reference internal" href="#id4">Usage</a></li>
<li><a class="reference internal" href="#example-1-fasta-as-input">Example-1: FASTA as input</a></li>
<li><a class="reference internal" href="#example-2-bed-as-input">Example-2: BED as input</a></li>
</ul>
</li>
<li><a class="reference internal" href="#use-cpat-to-detect-orf">Use CPAT to detect ORF</a><ul>
<li><a class="reference internal" href="#prepare-data">Prepare data</a></li>
<li><a class="reference internal" href="#run-cpat">Run CPAT</a></li>
<li><a class="reference internal" href="#check-the-results">Check the results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#how-to-choose-cutoff">How to choose cutoff</a></li>
<li><a class="reference internal" href="#how-to-prepare-training-dataset">How to prepare training dataset</a></li>
<li><a class="reference internal" href="#evaluating-performance">Evaluating Performance</a><ul>
<li><a class="reference internal" href="#figure-1">Figure-1</a></li>
<li><a class="reference internal" href="#figure-2">Figure-2</a></li>
<li><a class="reference internal" href="#id9">Figure-3</a></li>
</ul>
</li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#reference">Reference</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">CPAT</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Release history</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="_images/cpat.png"><img alt="alternate text" src="_images/cpat.png" style="width: 325.0px; height: 100.0px;" /></a>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This documentation is for CPAT v3.0.0 or future versions. For documentation of CPAT v2.0.0
and older versions, please go to <a class="reference external" href="http://rna-cpat.sourceforge.net/">http://rna-cpat.sourceforge.net/</a></p>
</div>
<section id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this heading"></a></h1>
<p><strong>CPAT v3.0.5</strong> (01/24/2024)</p>
<p>Use “pyproject.toml” to replace “setup.py”.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cpat.py</span></code> is now renamed to <code class="docutils literal notranslate"><span class="pre">cpat</span></code>. Similarly,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">make_hexamer_tab.py</span></code> is renamed to <code class="docutils literal notranslate"><span class="pre">make_hexamer_tab</span></code>,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">make_logitModel.py</span></code> is renamed to <code class="docutils literal notranslate"><span class="pre">make_logitModel</span></code>.</p></li>
<li><p>To ensure compatibility with the previous pipelines, please add the following lines into your ~/.bashrc file.</p></li>
</ul>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">alias</span><span class="w"> </span>cpat.py<span class="o">=</span><span class="s1">&#39;cpat&#39;</span>
<span class="nb">alias</span><span class="w"> </span>make_hexamer_tab.py<span class="o">=</span><span class="s1">&#39;make_hexamer_tab&#39;</span>
<span class="nb">alias</span><span class="w"> </span>make_logitModel.py<span class="o">=</span><span class="s1">&#39;make_logitModel&#39;</span>
</pre></div>
</div>
<p><strong>CPAT v3.0.4</strong> (05/26/2021)</p>
<p>Fix bug to read remote file for Python3.</p>
<p><strong>CPAT v3.0.3</strong> (03/08/2021)</p>
<p>Update “cpat.py” to handle alternative start codens.</p>
<p><strong>CPAT v3.0.2</strong> (08/17/2020)</p>
<p>Update “make_logitModel.py” to make it compatible with “cpat.py”.</p>
<p><strong>CPAT v3.0.1</strong></p>
<p>Minor bug fixed regarding the output format.</p>
<p><strong>CPAT v3.0.0</strong></p>
<p>For many transcripts, the longest ORF may not be the real ORF. For example, in human genome,
the 2nd longest ORF of NM_198086 is the real ORF, and the 3rd longest ORF of NM_030915 is the
real ORF. Version 3.0.0 is released to address this problem.</p>
<ol class="arabic simple">
<li><p>If model is provided, CPAT can be used as an ORFfinder. It gives exactly the same results as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a> does.</p></li>
<li><p>Search for all ORF candidates. The number of ORF reported is controlled by <code class="code docutils literal notranslate"><span class="pre">--min-orf</span></code> and <code class="code docutils literal notranslate"><span class="pre">--top-orf</span></code>.</p></li>
<li><p>In addition to basic ORF information (“ORF frame”, “ORF strand”, “ORF start”, “ORF end”, “ORF sequence”), it also reports “coding probability” for each ORF.</p></li>
<li><p>The best ORF will be selected (controlled by <code class="code docutils literal notranslate"><span class="pre">--best-orf</span></code>) either by <strong>ORF length</strong> or <strong>coding probability</strong>.</p></li>
</ol>
</section>
<section id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading"></a></h1>
<p>CPAT is a bioinformatics tool to predict RNA’s coding probability based on the RNA sequence characteristics. To achieve this goal,  CPAT calculates scores of these 4 linguistic features
from a set of known protein-coding genes and another set of non-coding genes.</p>
<ol class="arabic simple">
<li><p>ORF size</p></li>
<li><p>ORF coverage</p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/10/17/5303.abstract">Fickett TESTCODE</a></p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/20/24/6441.abstract">Hexamer usage bias</a></p></li>
</ol>
<p>CPAT will then builds a <a class="reference external" href="https://en.wikipedia.org/wiki/Logistic_regression">logistic regression</a> model using these 4 features as predictor variables and the “protein-coding status” as
the response variable. After evaluating the performance and determining the probability cutoff, the model can be used to predict the coding potential of new RNA sequences.</p>
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h1>
<section id="prerequisite">
<h2>Prerequisite<a class="headerlink" href="#prerequisite" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://www.python.org/download/releases/3.0/">python3.5</a> or later version</p></li>
<li><p><a class="reference external" href="http://numpy.scipy.org/">numpy</a></p></li>
<li><p><a class="reference external" href="https://pysam.readthedocs.io/en/latest/index.html">pysam</a></p></li>
<li><p><a class="reference external" href="http://www.r-project.org/">R</a></p></li>
</ol>
</section>
<section id="install-cpat-using-pip3">
<h2>install CPAT using pip3<a class="headerlink" href="#install-cpat-using-pip3" title="Permalink to this heading"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip3 install CPAT
$ pip3 install CPAT --upgrade  # if you already have CPAT v2.0 installed
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>User need to download prebuilt <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">logit model and hexamer table</a> for human, mouse, zebrafish and fly. For other species, we provide scripts to build these models (see below).</p></li>
</ul>
</div>
</section>
</section>
<section id="run-cpat-online">
<h1>Run CPAT online<a class="headerlink" href="#run-cpat-online" title="Permalink to this heading"></a></h1>
<p><a class="reference external" href="https://wlcb.oit.uci.edu/cpat/">https://wlcb.oit.uci.edu/cpat</a> is hosted by Dr Wei Li’s lab &#64; University of California Irvine.</p>
<p><strong>Step1: Upload data to CPAT server. There are 3 different ways to uploada</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Upload BED or FASTA format files from local disk. Files can be regular or compressed (.gz, .Z. .z, .bz, .bz2, .bzip2).</p></li>
<li><p>For small dataset, user can copy and paste data (in BED or FASTA format) directly to the text area.</p></li>
<li><p>For larger dataset, user can save data in web server (http, https or ftp) first, then paste the data url to text area. For very large dataset, run CPAT locally.</p></li>
</ul>
</div></blockquote>
<p><strong>Step2: Select Select Species assembly</strong></p>
<p><strong>Step3: Click Submit button</strong></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>This web server only supports Human (hg19), Mouse (mm9 and mm10), Fly (dm3) and Zebrafish (Zv9).</p></li>
<li><p>When input file is BED format, the reference genome is required and the assembly version is important.</p></li>
<li><p>When input file is FASTA format, the reference genome and the assembly version is ignored.</p></li>
</ul>
</div>
</section>
<section id="run-cpat-on-local-computer">
<h1>Run CPAT on local computer<a class="headerlink" href="#run-cpat-on-local-computer" title="Permalink to this heading"></a></h1>
<section id="input-files">
<h2>Input files<a class="headerlink" href="#input-files" title="Permalink to this heading"></a></h2>
<p>User needs to provide a gene file (‘-g’), a logit model file (‘-d’), a hexamer frequency table file (‘-x’)
and specify the output file name(‘-o’). Gene file could be either in <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1.7">BED</a>
or <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format. If in BED format, user also needs to specify the reference genome sequence file (‘-r’).</p>
</section>
<section id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">cpat</span> <span class="pre">-h</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>cpat  [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -g GENE_FILE, --gene=GENE_FILE
                        Genomic sequnence(s) of RNA in FASTA
                        (https://en.wikipedia.org/wiki/FASTA_format) or
                        standard 12-column  BED
                        (https://genome.ucsc.edu/FAQ/FAQformat.html#format1)
                        format. It is recommended to use *short* and *unique*
                        sequence identifiers (such as Ensembl transcript id)
                        in FASTA and BED file. If this is a BED file,
                        reference genome (&#39;-r/--ref&#39;) should be specified.
                        The input FASTA or BED file could be a regular text
                        file or compressed file (*.gz, *.bz2) or accessible
                        URL (http://, https://, ftp://). URL file cannot be a
                        compressed file.
  -o OUT_FILE, --outfile=OUT_FILE
                        The prefix of output files.
  -d LOGIT_MODEL, --logitModel=LOGIT_MODEL
                        Logistic regression model. The prebuilt models
                        for Human, Mouse, Fly, Zebrafish are availablel.
                        Run &#39;make_logitModel.py&#39; to build logistic
                        regression model for your own training datset.
  -x HEXAMER_DAT, --hex=HEXAMER_DAT
                        The hexamer frequency table.                       The
                        prebuilt tables for Human, Mouse, Fly, Zebrafish
                        are availablel. Run &#39;make_hexamer_tab.py&#39; to make this
                        table for your own training dataset.
  -r REF_GENOME, --ref=REF_GENOME
                        Reference genome sequences in FASTA format.
                        Reference genome file will be indexed automatically
                        if the index file ( *.fai) does not exist. Will be
                        ignored if FASTA file was provided to &#39;-g/--gene&#39;.
  --antisense           Logical to determine whether to search for ORFs
                        from the anti-sense strand. *Sense strand* (or coding
                        strand) is DNA strand that carries the translatable
                        code in the 5′ to 3′ direction. default=False (i.e.
                        only search for ORFs from the sense strand)
  --start=START_CODONS  Start codon (use &#39;T&#39; instead of &#39;U&#39;) used to
                        define the start of open reading frame (ORF).
                        default=ATG
  --stop=STOP_CODONS    Stop codon (use &#39;T&#39; instead of &#39;U&#39;) used to
                        define the end of open reading frame (ORF). Multiple
                        stop codons are separated by &#39;,&#39;. default=TAG, TAA,
                        TGA
  --min-orf=MIN_ORF_LEN
                        Minimum ORF length in nucleotides.
                        default=75
  --top-orf=N_TOP_ORF   Number of ORF candidates reported. RNAs may
                        have dozens of putative ORFs, in most cases, the real
                        ORF is ranked (by size) in the top several. It is not
                        necessary to calculate &quot;Fickett score&quot;,
                        &quot;Hexamer score&quot; and &quot;coding probability&quot; for every
                        ORF. default=5
  --width=LINE_WIDTH    Line width of output ORFs in FASTA format.
                        default=100
  --log-file=LOG_FILE   Name of log file. default=&quot;CPAT_run_info.log&quot;
  --best-orf=MODE       Criteria to select the best ORF: &quot;l&quot;=length,
                        selection according to the &quot;ORF length&quot;;
                        &quot;p&quot;=probability, selection according to the
                        &quot;coding probability&quot;. default=&quot;p&quot;
  --verbose             Logical to determine if detailed running
                        information is printed to screen.
</pre></div>
</div>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this heading"></a></h2>
<p>Use local FASTA file as input</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">cpat</span> <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span> <span class="pre">--antisense</span> <span class="pre">-d</span> <span class="pre">Human_logitModel.RData</span> <span class="pre">--top-orf=5</span> <span class="pre">-g</span> <span class="pre">Human_test_coding_mRNA.fa</span> <span class="pre">-o</span> <span class="pre">output1</span></code></p>
<p>Use a remote FASTA file as input</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">cpat</span> <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span> <span class="pre">--antisense</span> <span class="pre">-d</span> <span class="pre">Human_logitModel.RData</span> <span class="pre">--top-orf=5</span> <span class="pre">-g</span> <span class="pre">https://data.cyverse.org/dav-anon/iplant/home/liguow/CPAT/Human_test_coding_mRNA.fa</span> <span class="pre">-o</span> <span class="pre">output2</span></code></p>
<p>Use BED file as input. ‘-r’ is required</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">cpat</span> <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span> <span class="pre">--antisense</span> <span class="pre">-d</span> <span class="pre">Human_logitModel.RData</span> <span class="pre">--top-orf=5</span> <span class="pre">-g</span> <span class="pre">Human_test_coding_mRNA_hg19.bed</span> <span class="pre">-r</span> <span class="pre">hg19.fa</span> <span class="pre">-o</span> <span class="pre">output3</span></code></p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading"></a></h2>
<dl class="simple glossary">
<dt id="term-1.-output.ORF_seqs.fa">1. output.ORF_seqs.fa<a class="headerlink" href="#term-1.-output.ORF_seqs.fa" title="Permalink to this term"></a></dt><dd><p>The top ORF sequences (at least 75 nucleotides long) in FASTA format.</p>
</dd>
<dt id="term-2.-output.ORF_prob.tsv">2. output.ORF_prob.tsv<a class="headerlink" href="#term-2.-output.ORF_prob.tsv" title="Permalink to this term"></a></dt><dd><p>ORF information (strand, frame, start, end, size, Fickett TESTCODE score, Hexamer score) and coding probability)</p>
</dd>
<dt id="term-3.-output.ORF_prob.best.tsv">3. output.ORF_prob.best.tsv<a class="headerlink" href="#term-3.-output.ORF_prob.best.tsv" title="Permalink to this term"></a></dt><dd><p>The information of the best ORF. This file is a subset of “output.ORF_prob.tsv”</p>
</dd>
<dt id="term-4.-output.no_ORF.txt">4. output.no_ORF.txt<a class="headerlink" href="#term-4.-output.no_ORF.txt" title="Permalink to this term"></a></dt><dd><p>Sequence IDs or BED entries with no ORF found. Should be considered as non-coding.</p>
</dd>
<dt id="term-5.-output1.r">5. output1.r<a class="headerlink" href="#term-5.-output1.r" title="Permalink to this term"></a></dt><dd><p>Rscript file.</p>
</dd>
<dt id="term-6.-CPAT_run_info.log">6. CPAT_run_info.log<a class="headerlink" href="#term-6.-CPAT_run_info.log" title="Permalink to this term"></a></dt><dd><p>The log file.</p>
</dd>
</dl>
</section>
</section>
<section id="build-your-own-hexamer-table">
<h1>Build your own hexamer table<a class="headerlink" href="#build-your-own-hexamer-table" title="Permalink to this heading"></a></h1>
<p><code class="docutils literal notranslate"><span class="pre">make_hexamer_tab</span></code> calculates the in frame hexamer (6mer) frequency from <a class="reference external" href="https://en.wikipedia.org/wiki/Coding_region">CDS</a> sequence in fasta format. A <a class="reference external" href="https://en.wikipedia.org/wiki/Coding_region">CDS</a> is an mRNA sequence without the 3’ <a class="reference external" href="https://en.wikipedia.org/wiki/Untranslated_region">UTR</a> and 5’ <a class="reference external" href="https://en.wikipedia.org/wiki/Untranslated_region">UTR</a> regions.
This table is required by <code class="docutils literal notranslate"><span class="pre">cpat</span></code> to calculate the hexamer usage score. Users can download prebuilt hexamer tables (Human, Mouse, Fly, Zebrafish) from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">here</a>.</p>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">make_hexamer_tab</span> <span class="pre">-h</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>make_hexamer_tab  [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -c CODING_FILE, --cod=CODING_FILE
                        Coding sequence (must be CDS without UTR, i.e.
                        from start coden to stop coden) in fasta format
  -n NONCODING_FILE, --noncod=NONCODING_FILE
                        Noncoding sequences in fasta format
</pre></div>
</div>
</section>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this heading"></a></h2>
<p>First, download these two files:</p>
<blockquote>
<div><ul class="simple">
<li><p>Coding CDS sequences in FASTA format: <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/test/Human_coding_transcripts_CDS.fa.gz">Human_coding_transcripts_CDS.fa.gz</a></p></li>
<li><p>Noncoding sequences in FASTA format: <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/test/Human_noncoding_transcripts_RNA.fa.gz">Human_noncoding_transcripts_RNA.fa.gz</a></p></li>
</ul>
</div></blockquote>
<p>Then, run:</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">make_hexamer_tab</span> <span class="pre">-c</span> <span class="pre">Human_coding_transcripts_CDS.fa.gz</span> <span class="pre">-n</span> <span class="pre">Human_noncoding_transcripts_RNA.fa.gz</span> <span class="pre">&gt;Human_Hexamer.tsv</span></code></p>
</section>
<section id="id3">
<h2>Output<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">head</span> <span class="pre">Human_Hexamer.tsv</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>hexamer coding  noncoding
AAAAAA  0.0006471106736092786 0.001606589931772997
AAAAAC  0.00042092373222007566  0.0005113004850646316
AAAAAG  0.0008133623112408557 0.0006870944872085282
AAAAAT  0.0005917287586530271 0.0009504638599970318
AAAACA  0.0004934602747535982 0.0007256901384894673
AAAACC  0.0004003805362324795 0.0003686803641407804
AAAACG  9.064420497619743e-05 0.00010448394168197091
AAAACT  0.0004068399947646618 0.0004784022870680216
AAAAGA  0.0004286539039061299 0.000774026596998453
...
</pre></div>
</div>
</section>
</section>
<section id="build-your-own-logit-model">
<h1>Build your own logit model<a class="headerlink" href="#build-your-own-logit-model" title="Permalink to this heading"></a></h1>
<p>Build logistic regression model (“prefix.logit.RData”) required by <code class="docutils literal notranslate"><span class="pre">cpat</span></code>. This program will output
3 files:</p>
<ul class="simple">
<li><p>prefix.feature.xls: A table contains features calculated from training datasets (coding and noncoding gene lists).</p></li>
<li><p>prefix.logit.RData: logit model required by CPAT (if R was installed).</p></li>
<li><p>prefix.make_logitModel.r: R script to build the above logit model.</p></li>
</ul>
<p>Note: Users can download <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model">prebuilt logit models</a> for Human, Mouse, Fly and Zebrafish.</p>
<section id="id4">
<h2>Usage<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>make_logitModel  [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -c CODING_FILE, --cgene=CODING_FILE
                        Genomic sequnences of protein-coding RNAs in
                        FASTA or standard 12-column BED format. It is
                        recommended to use *short* and *unique* sequence
                        identifiers (such as Ensembl transcript id) in
                        FASTA and BED file. The input FASTA or BED file
                        could be a regular text file or compressed file
                        (*.gz, *.bz2) or accessible URL (http://, https://,
                        ftp://). When BED file is provided, use the ORF
                        defined in the BED file (the 7th and 8th columns in
                        BED file define the positions of &#39;start codon,
                        and &#39;stop codon&#39;, respectively). When FASTA file
                        is provided, searching for the longet ORF. For
                        well annotated genome, we recommend using BED file
                        as input because the longest ORF predicted from
                        RNA sequence might not be the real ORF. If this is
                        a BED file, reference genome (&#39;-r/--ref&#39;) should be
                        specified.
  -n NONCODING_FILE, --ngene=NONCODING_FILE
                        Genomic sequences of non-coding RNAs in FASTA or
                        standard 12-column BED format. It is recommended to
                        use *short* and *unique* sequence identifiers (such as
                        Ensembl transcript id) in FASTA and BED file. The
                        input FASTA or BED file could be a regular text file
                        or compressed file (*.gz, *.bz2) or accessible URL
                        (http://, https://, ftp://). If this is a BED file,
                        reference genome (&#39;-r/--ref&#39;) should be specified.
  -o OUT_FILE, --outfile=OUT_FILE
                        The prefix of output files.
  -x HEXAMER_DAT, --hex=HEXAMER_DAT
                        Hexamer frequency table. CPAT has prebuilt hexamer
                        frequency tables for Human, Mouse, Fly, Zebrafish. Run
                        &#39;make_hexamer_tab.py&#39; to generate this table.
  -r REF_GENOME, --ref=REF_GENOME
                        Reference genome sequences in FASTA format.
                        Ignore this option if mRNA sequences file was provided
                        to &#39;-g&#39;. Reference genome file will be indexed
                        automatically if the index file  *.fai) does not
                        exist.
  -s START_CODONS, --start=START_CODONS
                        Start codon (use &#39;T&#39; instead of &#39;U&#39;) used to
                        define the start of open reading frame (ORF).
                        default=ATG
  -t STOP_CODONS, --stop=STOP_CODONS
                        Stop codon (use &#39;T&#39; instead of &#39;U&#39;) used to
                        define the end of open reading frame (ORF).
                        Multiple stop codons are separated by &#39;,&#39;.
                        default=TAG, TAA, TGA
  --min-orf=MIN_ORF_LEN
                        Minimum ORF length in nucleotides.
                        default=30
  --log-file=LOG_FILE   Name of log file.
                        default=&quot;make_logitModel_run_info.log&quot;
  --verbose             Logical to determine if detailed running
                        information is printed to screen.
</pre></div>
</div>
</section>
<section id="example-1-fasta-as-input">
<h2>Example-1: FASTA as input<a class="headerlink" href="#example-1-fasta-as-input" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">make_logitModel</span> <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span> <span class="pre">-c</span> <span class="pre">Human_coding_transcripts_mRNA.fa.gz</span> <span class="pre">-n</span> <span class="pre">Human_noncoding_transcripts_RNA.fa.gz</span> <span class="pre">-o</span> <span class="pre">Human</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>2024-01-25 10:41:34 [INFO]  Start codons used: &quot;ATG&quot;
2024-01-25 10:41:34 [INFO]  Stop codons used: &quot;TAG, TAA, TGA&quot;
2024-01-25 10:41:34 [INFO]  Reading hexamer frequency table file: &quot;Human_Hexamer.tsv&quot;
2024-01-25 10:41:34 [INFO]  Process protein-coding RNA file: &quot;Human_coding_transcripts_mRNA.fa.gz&quot;
2024-01-25 10:41:34 [INFO]  Protein-coding RNA file &quot;Human_coding_transcripts_mRNA.fa.gz&quot; is in FASTA format
2024-01-25 10:42:12 [INFO]  Total 17984 coding sequences finished.
2024-01-25 10:42:12 [INFO]  Process non-coding RNA file: &quot;Human_noncoding_transcripts_RNA.fa.gz&quot;
2024-01-25 10:42:12 [INFO]  Non-coding RNA file &quot;Human_noncoding_transcripts_RNA.fa.gz&quot; is in FASTA format
2024-01-25 10:42:20 [INFO]  Total 11519 non-coding sequences finished.
2024-01-25 10:42:20 [INFO]  Wrting to &quot;Human.feature.xls&quot;
2024-01-25 10:42:20 [INFO]  Making logistic regression model from &quot;Human.feature.xls&quot; ...
...
</pre></div>
</div>
</section>
<section id="example-2-bed-as-input">
<h2>Example-2: BED as input<a class="headerlink" href="#example-2-bed-as-input" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">make_logitModel</span>&#160; <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span> <span class="pre">-c</span> <span class="pre">Human_coding_transcripts_hg19.bed</span> <span class="pre">-n</span> <span class="pre">Human_noncoding_transcripts_hg19.bed</span>&#160; <span class="pre">-r</span> <span class="pre">/database/hg19.fa</span>&#160; <span class="pre">-o</span> <span class="pre">Human</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>2024-01-25 10:44:45 [INFO]  Start codons used: &quot;ATG&quot;
2024-01-25 10:44:45 [INFO]  Stop codons used: &quot;TAG, TAA, TGA&quot;
2024-01-25 10:44:45 [INFO]  Reading hexamer frequency table file: &quot;Human_Hexamer.tsv&quot;
2024-01-25 10:44:45 [INFO]  Process protein-coding RNA file: &quot;Human_coding_transcripts_hg19.bed&quot;
2024-01-25 10:44:45 [INFO]  Protein-coding RNA file &quot;Human_coding_transcripts_hg19.bed&quot; is in BED format
...
</pre></div>
</div>
</section>
</section>
<section id="use-cpat-to-detect-orf">
<h1>Use CPAT to detect ORF<a class="headerlink" href="#use-cpat-to-detect-orf" title="Permalink to this heading"></a></h1>
<p>When using CPAT to find ORFs, it will gives exactly the same results as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a>.</p>
<section id="prepare-data">
<h2>Prepare data<a class="headerlink" href="#prepare-data" title="Permalink to this heading"></a></h2>
<p>Below is the mRNA sequence of protein-coding gene <a class="reference external" href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=UQCR10">UQCR10</a>. Copy and save it as “test.fa”.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;NM_013387.4
GCGGTGGCGCGAGTTGGACTGTGAAGAAACATGGCGGCCGCGACGTTGAC
TTCGAAATTGTACTCCCTGCTGTTCCGCAGGACCTCCACCTTCGCCCTCA
CCATCATCGTGGGCGTCATGTTCTTCGAGCGCGCCTTCGATCAAGGCGCG
GACGCTATCTACGACCACATCAACGAGGGGAAGCTGTGGAAACACATCAA
GCACAAGTATGAGAACAAGTAGTTCCTTGGAGGCCCCCATCCAGGCCAGA
AGGACCAGGTCCACCCAGCAGCTGTTTGCCCAGAGCTGGAGCCTCAGCTT
GAAGATGATGCTCAAGGTACTCTTCATGGACCACCATTCGCTGTTGGCAA
GAAACGGCTTTACTTACAAAACAGACTCTTTACCTTCTGCTGTGTTTGAA
GTATGTTTAGTCAGCATGCTCAGGAAATAAATGTGAATTGCCCTTGAGAC
CTGCTTCTACATTGGTTGCTTTGTTAACTCTACCTGATCTTCACTTGTCA
GTAATTTGAGACCACTTCAAAGCCCTCTGCAAACACCCCAAAGGCAGAAT
CTGCTATTTTGAGTTTTCCATTAACTTCCAAAGAATTCTGGTTTTCAAAA
CAGGAGCCAGAGTTGGAGATATTACAGTCAACTTTGGCTTCTAAGCCAGT
AATTCCATTCTTAAATACCTCACTGTCTTGGCCATGGGGAAGCACTATGG
CCTCAGCTGGGGGAAAGACCCTGGCCTAGGGGTCTTAGCCACTCCCCACC
CTAGGGTATAGTTCAGGGGTATCCAATCCTTTGGCTTCCCTGGGCCATGT
TGGAAGAATTGTCTTGGGCCACACATAAAATACAGTAACCATAGCTGATG
AGCTAAAACAAAAAACAATGGTTTGTGCAAAAATCTCATAATGTTTTAAT
AAAGTTGAAGAATTTG
</pre></div>
</div>
</section>
<section id="run-cpat">
<h2>Run CPAT<a class="headerlink" href="#run-cpat" title="Permalink to this heading"></a></h2>
<p>The command to run cpat.py is as below:</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">cpat</span> <span class="pre">-x</span> <span class="pre">Human_Hexamer.tsv</span>&#160; <span class="pre">-d</span>&#160; <span class="pre">Human_logitModel.RData</span>&#160; <span class="pre">--top-orf=100</span>&#160; <span class="pre">--antisense</span> <span class="pre">-g</span> <span class="pre">test.fa</span> <span class="pre">-o</span> <span class="pre">output</span></code></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>You must specify <code class="code docutils literal notranslate"><span class="pre">--antisense</span></code>, otherwise, it will only search ORFs from the sense strand.</p></li>
<li><p>You also specify <code class="code docutils literal notranslate"><span class="pre">--top-orf</span></code> to a large number to report all the ORFs.</p></li>
<li><p>The <code class="code docutils literal notranslate"><span class="pre">--min-orf</span></code> is set to 75 by default, same as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a>.</p></li>
</ul>
</div>
</section>
<section id="check-the-results">
<h2>Check the results<a class="headerlink" href="#check-the-results" title="Permalink to this heading"></a></h2>
<p>A total of 8 ORFs were found (sorted by the ORF size, the 7th column). If you copy and paste the same sequence
to <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a> web server, you will get <strong>exactly the same</strong> results.</p>
<p><code class="docutils literal notranslate"><span class="pre">$cat</span> <span class="pre">output.ORF_prob.tsv</span></code></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>ID      mRNA    ORF_strand      ORF_frame       ORF_start       ORF_end ORF     Fickett Hexamer Coding_prob
NM_013387.4_ORF_1       916     -       2       327     1       327     1.103   0.28998918917275        0.792763525921043
NM_013387.4_ORF_2       916     +       2       209     430     222     1.1605  0.0674464550896935      0.271842476390681
NM_013387.4_ORF_3       916     -       1       889     695     195     0.9192  -0.32000518247443       0.0113140534730678
NM_013387.4_ORF_4       916     +       1       31      222     192     1.2952  0.600469985268255       0.915129459422605
NM_013387.4_ORF_5       916     -       1       337     197     141     1.1626  0.133867810597757       0.185245402415541
NM_013387.4_ORF_6       916     -       3       119     3       117     1.2673  0.442351820001225       0.618496534888714
NM_013387.4_ORF_7       916     -       3       842     735     108     0.5832  -0.19401829042094       0.00290794398512764
NM_013387.4_ORF_8       916     +       3       684     761     78      0.7415  -0.154613060436537      0.00454929869181486
</pre></div>
</div>
<p>CPAT offers <strong>Fickett’s TESTCODE score</strong>, <strong>Hexamer score</strong>, and <strong>coding probability</strong> for each ORF. While the largest ORF is often the most probable for most mRNAs, it’s not always the case. In the instance of NM_013387.4, the ORF_4 is the most likely to code for protein, evident from its highest coding probability, despite not being the largest ORF. This is confirmed through BLATing the 8 ORF sequences to the reference genome.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/UQCR10.jpg"><img alt="_images/UQCR10.jpg" src="_images/UQCR10.jpg" style="width: 750.0px; height: 135.0px;" /></a>
</figure>
</section>
</section>
<section id="how-to-choose-cutoff">
<h1>How to choose cutoff<a class="headerlink" href="#how-to-choose-cutoff" title="Permalink to this heading"></a></h1>
<p><strong>Optimum cutoffs were determined from TG-ROC</strong>. For example, in human, coding probability (CP) cutoff &gt;=0.364 indicates coding sequence, CP &lt; 0.364 indicates noncoding sequence.</p>
<table class="docutils align-default" id="id10">
<caption><span class="caption-text">Optimum cutoffs</span><a class="headerlink" href="#id10" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 55.6%" />
<col style="width: 44.4%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Species</p></th>
<th class="head"><p>Coding probability threshold</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Human (<em>Homo sapiens</em>)</p></td>
<td><p>0.364</p></td>
</tr>
<tr class="row-odd"><td><p>Mouse (<em>Mus musculus</em>)</p></td>
<td><p>0.44</p></td>
</tr>
<tr class="row-even"><td><p>Fly (<em>Drosophila melanogaster</em>)</p></td>
<td><p>0.39</p></td>
</tr>
<tr class="row-odd"><td><p>Zebrafish (<em>Danio rerio</em>)</p></td>
<td><p>0.38</p></td>
</tr>
</tbody>
</table>
<p><strong>Here we provide the R code and the data that we used to generate</strong> <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74/F3.expansion.html">Figure 3</a> <strong>in our paper</strong>. <strong>Note the</strong> <a class="reference external" href="http://cran.r-project.org/web/packages/ROCR/index.html">ROCR</a> <strong>library is required to run our R code</strong>.</p>
<p>1) Download R code and data from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/Figure3_data/">here</a></p>
<p>2) Put the R code and the data table in the same folder</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls
10Fold_CrossValidation.r       Human_train.dat
</pre></div>
</div>
<p>3) Run the R code from command line or console. The R code will perform 10-fold cross validation and generate Figure_3.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ Rscript 10Fold_CrossValidation.r     # install ROCR before running this code

Loading required package: gplots
Attaching package: ‘gplots’
The following object is masked from ‘package:stats’:

    lowess

Loading required package: methods
Warning message:
package ‘gplots’ was built under R version 3.1.2
[1] &quot;ID&quot;      &quot;mRNA&quot;    &quot;ORF&quot;     &quot;Fickett&quot; &quot;Hexamer&quot; &quot;Label&quot;
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
null device
          1
</pre></div>
</div>
</section>
<section id="how-to-prepare-training-dataset">
<h1>How to prepare training dataset<a class="headerlink" href="#how-to-prepare-training-dataset" title="Permalink to this heading"></a></h1>
<p>We prebuild hexamer tables and logit models for <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">human, mouse, fly and zebrafish</a>.
If you want to run CPAT for other species, you need to prepare your own training data.</p>
<blockquote>
<div><ul class="simple">
<li><p>Optimal training datasets exhibit balance, where the count of coding sequences is approximately equal to that of noncoding sequences.</p></li>
<li><p>In cases where the genome of your target species lacks sufficient annotation of ‘coding’ and ‘noncoding’ genes for constructing a training dataset, consider leveraging data from evolutionarily related species to build your model.</p></li>
</ul>
</div></blockquote>
</section>
<section id="evaluating-performance">
<h1>Evaluating Performance<a class="headerlink" href="#evaluating-performance" title="Permalink to this heading"></a></h1>
<section id="figure-1">
<h2>Figure-1<a class="headerlink" href="#figure-1" title="Permalink to this heading"></a></h2>
<p>Combinatorial effects of 3 major features. 10,000 coding genes (red dots) and 10,000 noncoding genes (blue dots) are clearly separated into two clusters. (below figure)</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/Figure_1A_features_3D.png"><img alt="_images/Figure_1A_features_3D.png" src="_images/Figure_1A_features_3D.png" style="width: 600.0px; height: 600.0px;" /></a>
</figure>
</section>
<section id="figure-2">
<h2>Figure-2<a class="headerlink" href="#figure-2" title="Permalink to this heading"></a></h2>
<p>Performance evaluation was conducted using 10-fold cross-validation, considering a dataset comprising 10,000 coding genes and 10,000 noncoding genes. The blue dotted curves depict the results of individual 10-fold cross-validations, while the red solid curve represents the averaged curve across 10 validation runs. The evaluation metrics include: (A) ROC curve, (B) Precision-Recall (PR) curve, (C) Accuracy plotted against cutoff values, and (D) Two graphical ROC curves aimed at determining the optimum cutoff value.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/CPAT_performance.png"><img alt="_images/CPAT_performance.png" src="_images/CPAT_performance.png" style="width: 600.0px; height: 600.0px;" /></a>
</figure>
</section>
<section id="id9">
<h2>Figure-3<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h2>
<p>To compare CPAT with CPC and PhyloCSF, we build an independent testing dataset that composed
of 4,000 high quality protein coding genes from Refseq annotation and 4,000 lincRNAs from
Human lincRNA catalog (Cabili et al., 2011). All 8000 genes were not included in the training
dataset of CPAT.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/Figure_4.png"><img alt="_images/Figure_4.png" src="_images/Figure_4.png" style="width: 600.0px; height: 600.0px;" /></a>
</figure>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/Figure_S2.png"><img alt="_images/Figure_S2.png" src="_images/Figure_S2.png" style="width: 600.0px; height: 600.0px;" /></a>
</figure>
</section>
</section>
<section id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this heading"></a></h1>
<p>CPAT is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</section>
<section id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this heading"></a></h1>
<p>Wang, L., Park, H. J., Dasari, S., Wang, S., Kocher, J.-P., &amp; Li, W. (2013). CPAT: Coding-Potential Assessment Tool using an alignment-free logistic regression model. Nucleic Acids Research, 41(6), e74. <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74">doi:10.1093/nar/gkt006</a></p>
</section>
<section id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p>Liguo Wang: wang.liguo AT mayo.edu</p></li>
<li><p>Wei Li: wei.li AT uci.edu</p></li>
</ul>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>